src/supermarket/app/models/cookbook_upload/archive.rb

```diff
@@ -78,26 +81,41 @@ def read(path)
     # @raise [NoPath] if the source has no path
     # @raise [Error] if the source is not a compatible archive
     #
-    # @yieldparam [::Gem::Package::TarReader::Entry] entry
+    # @yieldparam [::Archive::Entry] entry
+    # @yieldparam [::String] contents
     #
     # @example
     #   archive = CookbookUpload::Archive.new(tarball)
     #   archive.each do |entry|
     #     puts "#{entry.full_name} has the following content:\n#{entry.read}"
     #   end
     #
-    def each
+    def each_entry
       raise NoPath unless @source.respond_to?(:path)
+      raise NotGzipped unless gzipped?
 
       begin
-        Zlib::GzipReader.open(@source.path) do |gzip|
-          Gem::Package::TarReader.new(gzip) do |tar|
-            tar.each { |entry| yield entry }
+        ::Archive::Reader.open_filename(@source.path) do |tar|
+          tar.each_entry_with_data do |entry, data|
+            contents = data.is_a?(String) ? data : ''
+            yield entry, contents
           end
         end
-      rescue Zlib::GzipFile::Error => e
-        raise Error, e.message
+      rescue ::Archive::Error => e
+        case e.message
```

Comment:
Should add an `else` here to reraise the error if it comes back with a different message so we don't silently swallow it.

Reply:
Good catch. That's very true. Just say no to error swallowing.

Reply:
[Added an else](https://github.com/chef/supermarket/pull/1810/commits/4507051cc8bf972fe817462c81652a6d8128de7c#diff-b112e427d9f10480a410ca9caf8081ccR108).